File: backend/app/schemas/pay_voucher.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field


# ── GL 계정과목 ──


class GlAccountItem(BaseModel):
    id: int
    code: str
    name: str
    account_type: str
    is_net_pay_account: bool
    is_cash_account: bool
    is_active: bool
    sort_order: int
    created_at: datetime
    updated_at: datetime


class GlAccountListResponse(BaseModel):
    items: list[GlAccountItem]
    total_count: int


class GlAccountBatchItem(BaseModel):
    id: int | None = None
    code: str = Field(min_length=1, max_length=20)
    name: str = Field(min_length=1, max_length=100)
    account_type: str = Field(pattern="^(expense|liability|asset|equity|revenue)$")
    is_net_pay_account: bool = False
    is_cash_account: bool = False
    is_active: bool = True
    sort_order: int = 0


class GlAccountBatchRequest(BaseModel):
    items: list[GlAccountBatchItem]
    delete_ids: list[int] = []


class GlAccountBatchResponse(BaseModel):
    items: list[GlAccountItem]
    total_count: int
    inserted_count: int
    updated_count: int
    deleted_count: int


# ── 급여항목 → 계정 매핑 ──


class PayGlMappingItem(BaseModel):
    id: int
    pay_item_code: str
    pay_item_name: str | None = None
    gl_account_code: str
    gl_account_name: str | None = None
    effective_from: date
    note: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class PayGlMappingListResponse(BaseModel):
    items: list[PayGlMappingItem]
    total_count: int


class PayGlMappingBatchItem(BaseModel):
    id: int | None = None
    pay_item_code: str = Field(min_length=1, max_length=30)
    gl_account_code: str = Field(min_length=1, max_length=20)
    effective_from: date
    note: str | None = Field(default=None, max_length=200)
    is_active: bool = True


class PayGlMappingBatchRequest(BaseModel):
    items: list[PayGlMappingBatchItem]
    delete_ids: list[int] = []


class PayGlMappingBatchResponse(BaseModel):
    items: list[PayGlMappingItem]
    total_count: int
    inserted_count: int
    updated_count: int
    deleted_count: int


# ── 전표 ──


class PayVoucherLineItem(BaseModel):
    id: int
    line_no: int
    gl_account_code: str
    gl_account_name: str | None = None
    cost_center_code: str | None = None
    debit_amount: float
    credit_amount: float
    summary: str | None = None
    source_item_code: str | None = None
    created_at: datetime


class PayVoucherItem(BaseModel):
    id: int
    voucher_no: str
    run_id: int
    voucher_type: str
    year_month: str | None = None
    voucher_date: date
    status: str
    total_debit: float
    total_credit: float
    summary: str | None = None
    created_by: int | None = None
    confirmed_by: int | None = None
    confirmed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class PayVoucherListResponse(BaseModel):
    items: list[PayVoucherItem]
    total_count: int


class PayVoucherDetailResponse(BaseModel):
    voucher: PayVoucherItem
    lines: list[PayVoucherLineItem]


class PayVoucherGenerateRequest(BaseModel):
    run_id: int


class PayVoucherGenerateDisbursementRequest(BaseModel):
    run_id: int


class PayVoucherActionResponse(BaseModel):
    voucher: PayVoucherItem


class MappingGapItem(BaseModel):
    item_code: str
    item_name: str | None = None
    direction: str


class MappingGapsResponse(BaseModel):
    run_id: int
    missing_item_codes: list[MappingGapItem]

File: frontend/src/types/pay-voucher.ts
// ── GL 계정과목 ──
export type GlAccountItem = {
  id: number;
  code: string;
  name: string;
  account_type: "expense" | "liability" | "asset" | "equity" | "revenue";
  is_net_pay_account: boolean;
  is_active: boolean;
  sort_order: number;
  created_at: string;
  updated_at: string;
};

export type GlAccountListResponse = {
  items: GlAccountItem[];
  total_count: number;
};

export type GlAccountBatchItem = {
  id?: number | null;
  code: string;
  name: string;
  account_type: string;
  is_net_pay_account: boolean;
  is_active: boolean;
  sort_order: number;
};

export type GlAccountBatchRequest = {
  items: GlAccountBatchItem[];
  delete_ids?: number[];
};

export type GlAccountBatchResponse = {
  items: GlAccountItem[];
  total_count: number;
  inserted_count: number;
  updated_count: number;
  deleted_count: number;
};

// ── 급여항목-계정 매핑 ──
export type PayGlMappingItem = {
  id: number;
  pay_item_code: string;
  pay_item_name?: string | null;
  gl_account_code: string;
  gl_account_name?: string | null;
  effective_from: string;
  note?: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

export type PayGlMappingListResponse = {
  items: PayGlMappingItem[];
  total_count: number;
};

export type PayGlMappingBatchItem = {
  id?: number | null;
  pay_item_code: string;
  gl_account_code: string;
  effective_from: string;
  note?: string | null;
  is_active: boolean;
};

export type PayGlMappingBatchRequest = {
  items: PayGlMappingBatchItem[];
  delete_ids?: number[];
};

export type PayGlMappingBatchResponse = {
  items: PayGlMappingItem[];
  total_count: number;
  inserted_count: number;
  updated_count: number;
  deleted_count: number;
};

// ── 급여 전표 ──
export type PayVoucherItem = {
  id: number;
  voucher_no: string;
  run_id: number;
  year_month?: string | null;
  voucher_date: string;
  status: string;
  total_debit: number;
  total_credit: number;
  summary?: string | null;
  created_by?: number | null;
  confirmed_by?: number | null;
  confirmed_at?: string | null;
  created_at: string;
  updated_at: string;
};

export type PayVoucherListResponse = {
  items: PayVoucherItem[];
  total_count: number;
};

export type PayVoucherLineItem = {
  id: number;
  line_no: number;
  gl_account_code: string;
  gl_account_name?: string | null;
  cost_center_code?: string | null;
  debit_amount: number;
  credit_amount: number;
  summary?: string | null;
  source_item_code?: string | null;
  created_at: string;
};

export type PayVoucherDetailResponse = {
  voucher: PayVoucherItem;
  lines: PayVoucherLineItem[];
};

export type PayVoucherActionResponse = {
  voucher: PayVoucherItem;
};

export type MappingGapItem = {
  item_code: string;
  item_name?: string | null;
  direction: string;
};

export type MappingGapsResponse = {
  run_id: number;
  missing_item_codes: MappingGapItem[];
};
